### PR TITLE
fix(ipeps): Arnoldi precheck + sigma gauge fixes for implicit AD

### PR DESCRIPTION
**Pull request by yingjerkao**
## Summary

- Add Arnoldi spectral-radius precheck to implicit-AD backward — detects non-contractive J^T before Neumann/GMRES solve
- Fix two bugs in sigma gauge backward that prevented implicit AD from working:
  1. Backward used absolute Q matrices instead of relative sigma (Q_new @ Q_old†)
  2. Sigma matrices were differentiated through (30 power iterations + QR) instead of stop_gradient
- Add "direct" backward method (`ad_backward_method="direct"`) — skips fixed-point solve, like variPEPS's LINEAR_SOLVER default
- Fix SVD sign convention (`_fix_svd_signs`) to first-large-element (variPEPS `gauge_fixed_svd`)
- Fix `S^{-1/2}` NaN gradient trap in SVD projectors
- Optimizer catches `CTMRGGradientError` and triggers stall recovery

## Scope

**Validated on 1-site C4v** (D=2, χ=8). Implicit AD + sigma gauge reaches E = −0.6593 (50 steps), comparable to explicit AD (E = −0.6602).

For 2-site non-C4v: SVD projectors now produce finite gradients (previously NaN), but the direct backward drifts below the physical ground state. This matches the known #328 issue — full variPEPS-style optimization needs additional energy guards (future work).

## Key results

| Path | D | χ | E | Notes |
|------|---|---|---|-------|
| 1-site C4v, implicit VJP + sigma | 2 | 8 | −0.6593 | Works (was broken before) |
| 1-site C4v, direct backward | 2 | 8 | −0.6599 | Works, faster per step |
| 2-site non-C4v, direct + SVD proj | 2 | 8 | −1.49 | Finite grads but non-variational drift |
| variPEPS reference (2-site) | 2 | 8 | −0.6623 | Target |

## Changes

| File | Change |
|------|--------|
| `ad_utils.py` | `arnoldi_spectral_radius()`, `CTMRGGradientError`, Arnoldi precheck, relative sigma, stop_gradient sigma, "direct" backward, first-large-element `_fix_svd_signs` |
| `_ctm_projector.py` | Safe `S^{-1/2}` pattern (prevents NaN gradient from near-zero SVs) |
| `ipeps_config.py` | `adjoint_arnoldi_precheck`, `adjoint_arnoldi_threshold` fields |
| `ipeps_optimize.py` | try/except `CTMRGGradientError`, "direct" in config encoding |
| `ipeps_ad_paths.md` | Document Arnoldi precheck |
| `test_ad_utils.py` | Arnoldi tests, integration tests, config tests |
| `test_ipeps.py` | Optimizer recovery test |

## Test plan

- [x] 724 core tests pass
- [x] Arnoldi utility tests (identity, contraction, expansion, nonsymmetric)
- [x] Integration: raises on QR gauge (ρ >> 5), passes when disabled
- [x] Optimizer catches error and continues
- [x] Implicit AD converges with sigma gauge (1-site C4v)
- [x] SVD projectors produce finite gradients for 2-site
- [x] Element-wise CTM convergence with SVD projectors (diff=0 vs diff~1 for eigh)

### CODE REVIEW

**Comment by yingjerkao** [comment]
All three Codex review items are valid and should be addressed in a follow-up PR:

1. **Arnoldi threshold not serialized** — `adjoint_arnoldi_threshold` needs to be packed into `config_tuple` alongside `adjoint_arnoldi_precheck`, otherwise non-default thresholds are silently lost when `_ctm_tensor_converge_bwd` reconstructs the config. Quick fix: add one more field to the tuple packing in `_config_to_tuple` / `_config_from_tuple`.

2. **Complex dtype in Arnoldi buffers** — `Q = jnp.zeros((n, k+1))` and `H = jnp.zeros((k+1, k))` default to float64. For complex-valued iPEPS (e.g. spiral states, models with DM interaction), this silently drops imaginary parts and gives wrong spectral radius. Fix: `jnp.zeros(..., dtype=v0.dtype)`.

3. **Optax L-BFGS state not reset in 1-site recovery** — The `reset` branch rolls back params/env but doesn't reinitialize `opt_state`. The 2-site branch already handles this. Fix: add `opt_state = optimizer.init(params)` in the 1-site `reset` branch, matching the 2-site pattern.

None of these affect the current real-valued Heisenberg benchmarks, but (2) will bite when complex iPEPS models hit the Arnoldi path.

**Comment by yingjerkao** [comment]
All three review items addressed in 30db29b:

1. **Arnoldi threshold serialized** — `adjoint_arnoldi_threshold` now packed at index 14 in config tuple, round-trips correctly.
2. **Complex dtype preserved** — Arnoldi `Q` and `H` buffers use `v0.dtype`, so complex-valued iPEPS models get correct spectral radius.
3. **Optax L-BFGS state reset** — 1-site `reset` recovery branch now calls `opt_state = optimizer.init(params)`, matching the 2-site pattern.

724 core tests pass.

**Comment by yingjerkao** [comment]
**Update: 2-site non-C4v implicit AD now converges with SVD projectors**

VJP backward + SVD projectors + phase gauge + Hager-Zhang:
- 2-site D=2 χ=8: **E = −0.9055**, converged at step 18 (dE < 1e-8)
- No non-variational drift, no Arnoldi precheck trigger
- Neumann VJP converges (ρ(J^T) < 5 with SVD proj element-wise convergence)

Key finding: element-wise CTM convergence (SVD projectors, not eigh) is what makes the implicit backward stable. The `_fix_svd_signs` first-large-element convention + `S_rsqrt` NaN guard were both necessary.

Energy value needs normalization investigation vs variPEPS (−0.6623), but the optimization is stable and converging.

File: src/tenax/algorithms/ad_utils.py [hunk 1]
```diff
@@ -537,6 +552,7 @@ def _config_to_tuple(config) -> tuple:
             getattr(config, "forward_gauge", "qr"), 0
         ),
         _PB_STR_TO_INT.get(getattr(config, "projector_backward", "auto"), 0),
+        int(getattr(config, "adjoint_arnoldi_precheck", True)),
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/ad_utils.py:555]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Resolving as part of a full sweep of the unresolved review backlog — **390 threads across 203 PRs**, tracked in **#802**.

To be clear about what this does and does not mean: resolving is a cleanup of the review UI, **not** a statement that this comment was addressed. 390 stale threads on merged PRs were hiding live signal, which is exactly what had happened — the sweep surfaced several genuine defects that had been sitting unread, now filed as #792–#801 (among them an implicit-AD backward that discards its GMRES convergence flag, and a symmetry-core check that reports a nonconserving block as conserved).

Comments verified as still applicable were carried into those issues, with the verification stated. Comments I could not cheaply verify are listed in #802 rather than silently dropped — if this one was a P1 on `src/`, it is in that table.

Since this PR is merged, nothing here can be fixed by editing it; the issues are where the work lives now.
